compas_python_utils/cosmic_integration/totalMassEvolvedPerZ.py

```diff
@@ -1,198 +1,137 @@
-    #Mmin and Mmax have to be between x1 and x4
-    CDFmin = CDFbrokenPowerLaw(np.array([Mmin]), x1, x2, x3, x4, a1, a2, a3, C1)
-    CDFmax = CDFbrokenPowerLaw(np.array([Mmax]), x1, x2, x3, x4, a1, a2, a3, C1)
-    
-    #All the random drawing that we need
-    drawM1         = np.random.uniform(CDFmin,CDFmax,sampleSize)
-    drawBinary     = np.random.uniform(0,1,sampleSize)
-    drawM2         = np.random.uniform(0,1,sampleSize)
-
-    #All the arrays we want to fill
-    M1 = np.zeros(sampleSize)
-    M2 = np.zeros(sampleSize)
-
-    #Define the IMF broken powerlaw and calculate masses from inverted CDF
-
-
-    M1 = invertCDFbrokenPowerLaw(drawM1, x1, x2, x3, x4, a1, a2, a3, C1)
-
-    #Binary fraction is easier, since we draw between 0-1, every draw with 
-    #value above binary fraction = single star and every value below is binary
-    #for a single star we set M2=0 Msun.
-    #Note that we assume that the binary Fraction is mass indepent
-    #Future work to implenet Max Moe ps and qs options
-    maskBinary = drawBinary < binaryFraction  #booleans
-
-    #again for the secondary we assume the mass ratio distribution to be flat across
-    #the whole parameter range so then the drawM2 (if it is in a binary) 
-    #just becomes the mass fraction.
-
-    M2[maskBinary] = np.multiply(drawM2[maskBinary],M1[maskBinary])
-    #all the ones outside the mask remain zero
-    return M1, M2
-
-
-def inverseCDF(C, CDF, index, xmin, xmax):
-    #CDF sincle powerlaw
-    a =  (1./(index + 1)) * C * CDF**(index+1)
-    b =  (1./(index + 1)) * C * xmin**(index+1)
-    c =  (1./(index + 1)) * C * xmax**(index+1)
-    top    = ((CDF * (c-b))+b)*(index + 1)
-    bottom = C
-    return (top/bottom)**(1./(index+1))
-
+import numpy as np
+from scipy.integrate import quad
+import h5py as h5       # for reading in data
+
+
+def IMF(m, m1=0.01, m2=0.08, m3=0.5, m4=200.0, a12=0.3, a23=1.3, a34=2.3):
+    """Calculate the fraction of stellar mass between m and m + dm for a three part broken power law.
+    
+    Default values follow Kroupa (2001)
+            zeta(m) ~ m^(-a_ij)
+
+    Parameters
+    ----------
+    m : `float` or `np.ndarray`
+        Mass at which to evaluate
+    mi : float, optional
+        masses at which to transition the slope
+    aij : float, optional
+        slope of the IMF between mi and mj
+
+    Returns
+    -------
+    imf_vals
+        IMF evaluated at the given masses
+    """
+    # calculate normalisation constants that ensure the IMF is continuous
+    b1 = 1 / ( 
+                (m2**(1 - a12) - m1**(1 - a12)) / (1 - a12) \
+                + m2**(-(a12 - a23)) * (m3**(1 - a23) - m2**(1 - a23)) / (1 - a23) \
+                + m2**(-(a12 - a23)) * m3**(-(a23 - a34)) * (m4**(1 - a34) - m3**(1 - a34)) / (1 - a34)
+             )
+    b2 = b1 * m2**(-(a12 - a23))
+    b3 = b2 * m3**(-(a23 - a34))
+    
+    # evaluate IMF either at a point or for a list of points
+    if isinstance(m, float):
```

Comment:
If we put a @np.vectorize decorator around this function I think you can skip the float VS array logic. 

I'd be happy to make this change if you're not sure what I'm talking about!

Reply:
I made the suggested change

Reply:
This is magic, thanks for demonstrating this!